### DISCUSSION TITLE
Referencing in JavaScript with vite

### DISCUSSION THREAD

**Discussion by mickeyvip** [tailwindlabs/tailwindcss#3646] [Help] [answered]
Hello,

I am using `tailwindcss` with `vite` and the installation following the official documentation was smooth.

I also wanted to reference the theme from JavaScript and also followed the official docs (https://tailwindcss.com/docs/configuration#referencing-in-java-script).

Here I have problems.

There is a file `theme.js`:

```js
// theme.js
// --------

import resolveConfig from 'tailwindcss/resolveConfig'
import tailwindConfig from '../tailwind.config.js'

const fullConfig = resolveConfig(tailwindConfig)

export default fullConfig;
```

That throws an exception:

```
Uncaught SyntaxError: The requested module '/tailwind.config.js' does not provide an export named 'default'
```

If I try to import the `tailwindConfig` as:

```js
import * as tailwindConfig from '../tailwind.config.js'
```

There is another exception:

```
tailwind.config.js:1 Uncaught ReferenceError: module is not defined
    at tailwind.config.js:1
```

Or if there is something like:

```js
// theme.js
// --------

const defaultTheme = require('tailwindcss/defaultTheme');
// ...
```

the exception is:

```
tailwind.config.js:1 Uncaught ReferenceError: require is not defined
    at tailwind.config.js:1
```

Seems like the issue is that `tailwind.config.js` is a common js module.
Is the problem with `vite` or `tailwindcss` or am I doing something wrong?

Thank you for advance.

**Comment by schelmo** [comment, marked as answer]
adding it to optimizeDeps.include works for me:
```js
// vite.config.js
module.exports = {
  // ...
  resolve: {
    alias: {
      'tailwind.config.js': path.resolve(__dirname, 'tailwind.config.js'),
    },
  }
  optimizeDeps: {
    include: [
      path.resolve(__dirname, 'tailwind.config.js'),
    ]
  }
}
```
then you can import it with
```js
import config from 'tailwind.config.js'

**Reply by mickeyvip** [reply]
@schelmo , thank you for the response.

Just tried it locally on WSL2 and unfortunately it doesn't seems to work. I get the same error.

```js
// vite.config.js
import path from 'path';
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  resolve: {
    alias: {
      'tailwind.config.js': path.resolve(__dirname, 'tailwind.config.js'),
    },
  },
  optimizeDeps: {
    include: [
      path.resolve(__dirname, 'tailwind.config.js'),
    ]
  }
})
```

```js
// src/theme.js
import resolveConfig from 'tailwindcss/resolveConfig'
import tailwindConfig from '../tailwind.config.js'

export default resolveConfig(tailwindConfig)
```

```vue
// src/components/HelloWorld.vue
<template>
  <button @click="state.count++">count is: {{ state.count }}</button>
</template>

<script setup>
import { reactive } from 'vue'
import theme from '../theme'

console.log(theme)

const state = reactive({ count: 0 })
</script>

<style scoped>
</style>
```

```console
Uncaught SyntaxError: The requested module '/tailwind.config.js' does not provide an export named 'default'
```

**Reply by schelmo** [reply]
ah, change the optimizeDeps to:
```js
optimizeDeps: {
    include: [
      'tailwind.config.js',
    ]
  }
```
and use:
import tailwindConfig from 'tailwind.config.js'
without the relative ../

**Reply by mickeyvip** [reply]
Hi @schelmo.

Thank you for a quick response.

Changed to:

```js
// vite.config.js

import path from 'path'
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  resolve: {
    alias: {
      'tailwind.config.js': path.resolve(__dirname, 'tailwind.config.js'),
    },
  },
  optimizeDeps: {
    include: [
      'tailwind.config.js',
    ]
  }
})
```

I see that it adds `tailwind.config.js` as a dependency:

![image](https://user-images.githubusercontent.com/220427/109314158-ad4b2800-7851-11eb-85c5-4477ce46f0c1.png)

But it doesn't help. Still the same exception:

![image](https://user-images.githubusercontent.com/220427/109314244-c7850600-7851-11eb-98c9-f25e5fa6c228.png)

**Reply by schelmo** [reply]
did you adjust the import?
> and use:
> import tailwindConfig from 'tailwind.config.js'
> without the relative ../

**Reply by mickeyvip** [reply]
🤦🏻 

Now it works!

Thank you so much!

The documentation on Vite's website for config is very minimal and I would not know how to configure this without your help.

Thank you again!

**Reply by mickeyvip** [reply]
How do I mark this as "answered"?

**Reply by storm-monkey** [reply]
@schelmo 
It is working fine in dev serer.but it shows the same error on build.plz help me.
 ![image](https://user-images.githubusercontent.com/63580709/120010077-598bde80-bffa-11eb-82b9-85fb901dcb81.png)

**Reply by hivokas** [reply]
@Megatron-Siva did you have a chance to fix it? I'm facing the same issue.

**Reply by storm-monkey** [reply]
> @Megatron-Siva did you have a chance to fix it? I'm facing the same issue.

unfortunately no.

**Reply by hivokas** [reply]
@Megatron-Siva I've managed to fix it yesterday. Try this:

<img width="661" alt="image" src="https://user-images.githubusercontent.com/22997803/120839119-3bc9f680-c571-11eb-8358-0534acfa1725.png">

**Reply by jessedobbelaere** [reply]
Thanks @hivokas  that was the missing piece!

Note that the build dependencies of tailwind are still bundled along (as mentioned in the docs: https://tailwindcss.com/docs/configuration#referencing-in-java-script)

**Reply by donaldma** [reply]
@Megatron-Siva what fixed it for me was adding this plugin https://www.npmjs.com/package/@rollup/plugin-commonjs

**Reply by sbc640964** [reply]
> @Megatron-Siva I've managed to fix it yesterday. Try this:
> 
> <img alt="image" width="661" src="https://user-images.githubusercontent.com/22997803/120839119-3bc9f680-c571-11eb-8358-0534acfa1725.png">

I tried, it does not work in any way.

**Comment by couteau** [comment]
I am encountering this same issue. I've followed the instructions in this post, and I'm still getting there error. 

My setup:

```ts
// vite.config.ts
import path from 'path';
import { defineConfig } from "laravel-vite";
import vue from "@vitejs/plugin-vue";

export default defineConfig({
	plugins: [vue()],
	resolve: {
		alias: {
			'tailwind.config.js': path.resolve(__dirname, 'tailwind.config.js'),
		},
	},
	optimizeDeps: {
		include: [
			'tailwind.config.js',
		]
	}
})
```

```js
// tailwind.config.js
const colors = require('tailwindcss/colors')
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
    purge: [
        './vendor/laravel/framework/src/Illuminate/Pagination/resources/views/*.blade.php',
        './vendor/laravel/jetstream/**/*.blade.php',
        './storage/framework/views/*.php',
        './resources/views/**/*.blade.php',
        './resources/js/**/*.vue',
    ],

    darkMode: 'class',

    theme: {
        colors: {
            primary: {
                DEFAULT: '#20a6fb',
                '50': '#edf1f7',
                '100': '#d3e2f8',
                '200': '#b6d4f9',
                '300': '#97c6fa',
                '400': '#20a6fb',
                '500': '#4996db',
                '600': '#3085cb',
                '700': '#4b6fa9',
                '800': '#2e5786',
                '900': '#1F2E47',
            },
            transparent: 'transparent',
            current: 'currentColor',
            black: colors.black,
            white: colors.white,
            gray: colors.blueGray,
            blue: colors.blue,
            indigo: colors.indigo,
            green: colors.emerald,
            red: colors.rose,
            yellow: colors.amber,
        },
        extend: {
            fontFamily: {
                sans: ['Nunito', ...defaultTheme.fontFamily.sans],
            },
            transitionProperty: {
                height: 'min-height, max-height',
                font: 'font-size, font-weight',
            },
            spacing: {
                '120': '30rem',
                '144': '36rem',
            }
        },
    },

    variants: {
        extend: {
            opacity: ['disabled'],
            transitionProperty: ['hover'],
            textColor: ['disabled'],
            backgroundColor: ['disabled'],
        },
    },

    plugins: [require('@tailwindcss/forms')],
}
```

```js
// theme.js plugin
import resolveConfig from 'tailwindcss/resolveConfig'
import tailwindConfig from 'tailwind.config.js'

export default {
    install: (app, options) => {
        const theme = resolveConfig(tailwindConfig).theme
        app.config.globalProperties.$theme = theme
    }
}
```

The error message on 'yarn build': 

```
'default' is not exported by tailwind.config.js, imported by resources/plugins/theme.js
file: /Users/stuart/Source/test/resources/plugins/theme.js:3:7
1: import resolveConfig from 'tailwindcss/resolveConfig'
2: import tailwindConfig from 'tailwind.config.js'
          ^
3:
4: export default {
```

What am I doing wrong?

**Reply by richardtallent** [reply]
Can you `import foo` from a file that uses `module.exports = {}`? I'm a bit fuzzy on common JS vs. ES module details, but I usually see `module.exports = {}` paired with `require()`, and `export default {}` paired with `import`.

**Comment by schelmo** [comment]
i made a working example repository here:
https://github.com/schelmo/vite-use-tailwindconfig

**Reply by storm-monkey** [reply]
But ,it is not working in "npm run build"

**Reply by schelmo** [reply]
i pushed a fix (dont know if its the best way to go)

**Reply by johndwells** [reply]
@schelmo Thanks for creating that repo, it helped us overcome this issue!

**Reply by sbc640964** [reply]
> But ,it is not working in "npm run build"

For me too, what did you do in the end?

**Comment by TheoJL** [comment]
We fixed it by converting to ES6. 

`tailwind.config.js` we use `export default { ... }`. 

And all plugins are imported instead of required. 



In `postcss.config.js`
```
import tailwind from "tailwindcss";
import autoprefixer from "autoprefixer";
import tailwindConfig from "./tailwind.config.js";

export default {
    plugins: [tailwind(tailwindConfig), autoprefixer],
};
```

In `vite.config.js`

```
import postcss from "./postcss.config.js";
```
```
    css: {
        postcss,
    },
```

[https://stackoverflow.com/a/66406070/1356577](https://stackoverflow.com/a/66406070/1356577)

---------------------

The above solution worked great except for [Tailwind CSS Intellisense](https://github.com/tailwindlabs/tailwindcss-intellisense) for VSCode. 

We fixed it by renaming `tailwind.config.js` to `tailwind.config.mjs`.

**Reply by Huuums** [reply]
This worked perfectly for us too. 

Since we also wanted to have autocompletion we have 2 identical configs. 

One as `tailwind.config.mjs` which is used in postcss and the javascript file accessing the theme. 

Then the same config again as `tailwind.config.cjs` (with the needed adjustments like changing imports and exports to require and module.exports) so that the tailwind intellisense extension works. 

It's a dumb solution but it works until a proper fix is available.

**Reply by nariogulu** [reply]
> 

Hi, I have followed your instruction and converted everything to ES6 and thought it works, I get an annoying typescript error on postcss in vite config  

Type '{ plugins: (Plugin | Processor | typeof autoprefixer)[]; }' is not assignable to type 'string | (ProcessOptions & { plugins?: Plugin[] | undefined; }) | undefined'.
  Type '{ plugins: (Plugin | Processor | typeof autoprefixer)[]; }' is not assignable to type 'ProcessOptions & { plugins?: Plugin[] | undefined; }'.
    Type '{ plugins: (Plugin | Processor | typeof autoprefixer)[]; }' is not assignable to type '{ plugins?: Plugin[] | undefined; }'.
      Types of property 'plugins' are incompatible.
        Type '(Plugin | Processor | typeof autoprefixer)[]' is not assignable to type 'Plugin[]'.
          Type 'Plugin | Processor | typeof autoprefixer' is not assignable to type 'Plugin'.
            Property 'postcssPlugin' is missing in type 'Processor' but required in type 'Plugin'

I have tried everything possible to get rid of this error but to no avail. any help please?

**Reply by incutonez** [reply]
@nariogulu, I think it's because the export default in `postcss.config.ts` doesn't actually have a type... it's just an object.  I'm not entirely sure what the type should be... I tried `Processor`, but that didn't seem to work, so right now I'm just any'ing it... e.g.

```js
// postcss.config.ts
export default {
	plugins: [tailwind(tailwindConfig), autoprefixer],
} as any;
```

**Comment by fabianwohlfart** [comment]
Hi Guys,

Consider adding tailwind.config.js to commonjsOptions in the build property.
(Besides resolve and optimizeDeps)
For me this solved the issue.

### vite.config.js
```
build: {
  ...
  commonjsOptions: {
    include: ['tailwind.config.js', 'node_modules/**'],
  },
},
```

**Reply by kreierson** [reply]
This worked for me! Thank you sir!

**Comment by lobo-tuerto** [comment]
I wrote a blog post for those that want something to point people to when looking for answers for this: https://lobotuerto.com/notes/import-tailwind-config-in-vite

**Reply by mattanniah** [reply]
Legend!
Finally got this working with the help of your post. 
The key for me was the **vite-env.d.ts** file to make TypeScript not poop the bed. 

<img width="721" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/32303458/7f5e2338-3230-4400-aba8-c271186c4632">

I thought I might have to add 'tailwind-config' as a path in **tsconfig** but no, not necessary.

**Comment by Lauro235** [comment]
Is there a solution for this not in common js?

**Comment by Nagell** [comment]
Up. 
Fighting with this and still cannot find a way to run it with `.ts` and `satisfies Config`. It works nicely in the IDE by giving an infered type but sooner or later Vite and Storybook will throw some error.
